text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+
+```rust
+let yellow = { red: 0, green: 255, blue: 255 };
+
+let white = { red: 255, ..yellow };
+```
+
+To match the behaviour of FRU for nominal structs, we impose a restriction
+that the fields mentioned before `..yellow` must all exist in `yellow` and
+have the same types. This means that we cannot write:
+
+```rust
+let white_50_opacity = { alpha: 0.5, red: 255, ..yellow };
+```
+
+However, there is no fundamental reason why we could not allow this,
+but to start off more conservative and to be uniform in behaviour,
+we don't allow this for the time being.
+
+### The type of a structural record
+
+You might be wondering what the type of `my_record` that we've been using
+thus far is. Because this is structural typing, the fields are significant,
+so the type of the record is simply:
+
+```rust
+type TheRecord = { alpha: u8, beta: bool, gamma: &'static str };
+
+let my_record: TheRecord = { alpha: 42u8, beta: true, gamma: "Dancing Ferris" };
+```
+
+Notice how this matches the way we defined `MyRecordType` is we remove
+the prefix `struct MyRecordType`.
+
+The order in which we've put `alpha`, `beta`, and `gamma` here does not matter.
+We could have also written:
+
+```rust
+type TheRecord = { beta: bool, alpha: u8, gamma: &'static str };
+```
+
+As long as the type is a *permutation* of each pair of field name and field type,
+the type is the same. This also means that we can write: 
+
+```rust
+let my_record: TheRecord = { alpha: 42u8, gamma: "Dancing Ferris", beta: true };
+```
+
+### Implemented traits
+
+With respect to trait implementations, because the type is structural,
+and because there may be an unbound number of fields that can all be arbitrary
+identifiers, there's no way to define implementations for the usual traits
+in the language itself.
+
+[tuples]: https://doc.rust-lang.org/std/primitive.tuple.html
+
+Instead, the compiler will automatically provide trait implementations for the
+standard traits that are implemented for [tuples]. These traits are: `Clone`,
+`Copy`, `PartialEq`, `Eq`, `PartialOrd`, `Ord`, `Debug`, `Default`, and `Hash`.
+Each of these traits will only be implemented if all the field types of a struct
+implements the trait.
+
+For all of the aforementioned standard traits, the semantics of the
+implementations are similar to that of `#[derive(Trait)]` for named-field structs.
+
++ For cloning `{ alpha: A, beta: B, gamma: C }` the logic is simply:
+
+  ```rust
+  {
+      alpha: self.alpha.clone(),
+      beta: self.beta.clone(),
+      gamma: self.gamma.clone(),
+  }
+  ```
+
++ For `Default`, you would get:
+
+  ```rust
+  {
+      alpha: Default::default(),
+      beta: Default::default(),
+      gamma: Default::default(),
+  }
+  ```
+
++ For `PartialEq`, each field is compared with same field in `other: Self`.
+
++ For `ParialOrd` and `Ord`, lexicographic ordering is used based on
+  the name of the fields and not the order given because structural records
```

Comment:
With macros fields can have names with same textual representation, but different hygienic contexts.

Reply:
Does that change anything wrt. the implementations provided tho? I don't see any problems with hygiene intuitively, but maybe I can given elaboration?

Reply:
This is probably relevant to other places where the order of fields needs to be "normalized".

Reply:
Sure. :) But I'm not sure if you are pointing out a problem or just noting...
Any sorting / "normalization" is done post expansion.

Reply:
It's not clear to me how to sort hygienic contexts in stable order, especially in cross-crate scenarios. That probably can be figured out somehow though.